core/src/data_sources/splitter.rs

```diff
@@ -79,37 +81,111 @@ impl Splitter for BaseV0Splitter {
         embedder.initialize(credentials).await?;
 
         // Encode the clean text.
         let encoded = embedder.encode(clean).await?;
+        let mut decoded = vec![];
+
+        let mut start = 0;
+        while start < encoded.len() {
+            let end = cmp::min(start + max_chunk_size - 1, encoded.len() - 1);
 
-        // Split the encoded text into chunks of size max_chunk_size.
-        let mut chunks = Vec::new();
-        let mut chunk = Vec::new();
-        let mut chunk_size = 0;
-        for token in encoded {
-            chunk.push(token);
-            chunk_size += 1;
-            if chunk_size >= max_chunk_size {
-                chunks.push(chunk);
-                chunk = Vec::new();
-                chunk_size = 0;
+            match decode_chunk_with_remainder(&embedder, &encoded[start..=end]).await {
+                Ok((chunk_decoded, chunk_remainder)) => {
+                    
+                    if chunk_remainder.is_some() {
+                        let chunk_remainder_value = chunk_remainder.unwrap();
+                        if chunk_remainder_value.len() == max_chunk_size {
+                            return Err(anyhow!("The remainder chunk is the same size as the max chunk size - can't keep decoding. max_chunk_size: {}", max_chunk_size));
+                        }
+                        start += (end - start) + 1 - chunk_remainder_value.len();
+                    } else {
+                        start += (end - start) + 1;
+                    }
+                    
+                    if chunk_decoded.is_some() {
+                        decoded.push(chunk_decoded.unwrap());
+                    }
+                },
+                Err(e) => {
+                    return Err(e);
+                }
             }
         }
-        if chunk.len() > 0 {
-            chunks.push(chunk);
-        }
 
-        // Decode the chunks in parallel.
-        let mut futures = Vec::new();
-        for chunk in chunks {
-            futures.push(embedder.decode(chunk));
+        async fn decode_chunk_with_remainder(
+            embedder: &Box<dyn Embedder + Sync + Send>,
+            chunk: &[usize],
+        ) -> Result<(Option<String>, Option<Vec<usize>>), anyhow::Error> {
+            let mut remaining: Option<Vec<usize>> = None;
+            // The maximum number of tokens to slide the window by when decoding fails.
+            const MAX_ERROR_SLIDE : usize = 4;
+
+            let mut end = chunk.len();
+
+            while end > 0 {
+                if chunk.len() - end > MAX_ERROR_SLIDE {
+                    return Err(anyhow!("Failed to decode chunk after moving the sliding window one {} times by one.", MAX_ERROR_SLIDE));
+                }
+                match embedder.decode(chunk[0..end].to_vec()).await {
+                    Ok(decoded_token) => {
+
+                        if end != chunk.len() {
+                            remaining = Some(chunk[end..].to_vec());
```

Comment:
It's a big confusing to have remaining defined above and set here. You can just do something like:
```
let remaining = match (end != chunk.len()) {
  true => Some(chunk[end..].to_vec()),
  false => None
}
```

And not have it defied all the way above?